File: leetcode/medium/DesignCircularQueue.py
class MyCircularQueue:
    """
    A circular queue implementation using a fixed-size list.
    This design uses a size counter and modulus arithmetic for efficient
    and clean state management.
    """

    def __init__(self, k: int):
        # The underlying list to store queue elements.
        self.queue = [0] * k
        # The maximum number of elements the queue can hold.
        self.capacity = k
        # The number of elements currently in the queue.
        self.count = 0
        # `front` pointer: index of the first element in the queue.
        self.front = 0
        # `rear` pointer: index of the next available slot for a new element.
        self.rear = 0

    def enQueue(self, value: int) -> bool:
        """Inserts an element into the circular queue. Returns true if the operation is successful."""
        if self.isFull():
            return False
        
        # Place the new value at the rear pointer's location.
        self.queue[self.rear] = value
        # Move the rear pointer to the next available slot, wrapping around if necessary.
        self.rear = (self.rear + 1) % self.capacity
        # Increment the count of elements.
        self.count += 1
        return True

    def deQueue(self) -> bool:
        """Deletes an element from the circular queue. Returns true if the operation is successful."""
        if self.isEmpty():
            return False
        
        # Move the front pointer to the next element, effectively "removing" the current one.
        self.front = (self.front + 1) % self.capacity
        # Decrement the count of elements.
        self.count -= 1
        return True

    def Front(self) -> int:
        """Gets the front item from the queue."""
        if self.isEmpty():
            return -1
        # The front pointer always indicates the first element.
        return self.queue[self.front]

    def Rear(self) -> int:
        """Gets the last item from the queue."""
        if self.isEmpty():
            return -1
        # The last element is at the position just before the rear pointer.
        # We use modulus to handle the wrap-around case where rear is 0.
        last_element_index = (self.rear - 1 + self.capacity) % self.capacity
        return self.queue[last_element_index]

    def isEmpty(self) -> bool:
        """Checks whether the circular queue is empty or not."""
        return self.count == 0

    def isFull(self) -> bool:
        """Checks whether the circular queue is full or not."""
        return self.count == self.capacity

File: leetcode/medium/DesignCircularQueue.java
/**
 * Design your implementation of the circular queue. The circular queue is a linear data structure
 * in which the operations are performed based on FIFO (First In First Out) principle and the last position 
 * is connected back to the first position to make a circle. It is also called "Ring Buffer".
 * * The class uses an array to store elements and uses three variables:
 * - front: Index of the element at the front of the queue (if not empty).
 * - rear: Index of the next available slot for insertion.
 * - count: Current number of elements in the queue.
 * * All methods are O(1) time complexity.
 */
class MyCircularQueue {
    // Array to hold the queue elements.
    private final int[] queue;
    // Maximum capacity of the queue. Made final for clarity.
    private final int capacity; 
    // Current number of elements in the queue. Used to distinguish between empty and full.
    private int count;
    // Index of the element at the front of the queue.
    private int front;
    // Index of the next available slot (where the next element will be inserted).
    private int rear;

    /**
     * Initializes the object with the size of the queue to be k.
     * @param k The size of the circular queue.
     */
    public MyCircularQueue(int k) {
        // Use k+1 size for one common strategy, or use count/size for another.
        // This implementation uses 'count' and requires array size 'k'.
        queue = new int[k];
        capacity = k;
        count = 0;
        front = 0;
        rear = 0;
    }
    
    /**
     * Inserts an element into the circular queue. Return true if the operation is successful.
     * Time Complexity: O(1)
     * @param value The element to insert.
     * @return true if successful, false if the queue is full.
     */
    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }
        // Insert the new value at the 'rear' index.
        queue[rear] = value;
        
        // Move the 'rear' pointer to the next slot, wrapping around using modulo.
        rear = (rear + 1) % capacity;
        count++;
        return true;
    }
    
    /**
     * Deletes an element from the circular queue. Return true if the operation is successful.
     * Time Complexity: O(1)
     * @return true if successful, false if the queue is empty.
     */
    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }
        // Move the 'front' pointer to the next element, wrapping around using modulo.
        front = (front + 1) % capacity;
        count--;
        return true;
    }
    
    /**
     * Gets the front item from the queue. If the queue is empty, return -1.
     * Time Complexity: O(1)
     * @return The front element, or -1 if the queue is empty.
     */
    public int Front() {
        if (isEmpty()) {
            return -1;
        }
        // The element at the 'front' index is the current front of the queue.
        return queue[front];
    }
    
    /**
     * Gets the last item from the queue. If the queue is empty, return -1.
     * Time Complexity: O(1)
     * @return The rear element, or -1 if the queue is empty.
     */
    public int Rear() {
        if (isEmpty()) {
            return -1;
        }
        // The 'rear' pointer points to the *next empty slot*. 
        // The last inserted element is at the index before 'rear'.
        // Add 'capacity' before modulo to handle the case where rear is 0 (i.e., (0 - 1) % k).
        int lastElementIndex = (rear - 1 + capacity) % capacity;
        return queue[lastElementIndex];
    }
    
    /**
     * Checks whether the circular queue is empty or not.
     * Time Complexity: O(1)
     * @return true if the queue has no elements, false otherwise.
     */
    public boolean isEmpty() {
        return count == 0;
    }
    
    /**
     * Checks whether the circular queue is full or not.
     * Time Complexity: O(1)
     * @return true if the number of elements equals the capacity, false otherwise.
     */
    public boolean isFull() {
        return count == capacity;
    }
}